### PR TITLE
backport: merge bitcoin#24778, #24766, #24849, #24844, #24853, #24895, #24802, #24932, partial bitcoin#23212, #23462 (lint backports)

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

* Dependency for https://github.com/dashpay/dash/pull/6429

* Even though support for Python scripts was extended to `lint-all` in [bitcoin#24762](https://github.com/dashpay/dash/commit/f226e8dc1f4832f4ddf00a3bf82c23edb68fbedd) ([dash#6023](https://github.com/dashpay/dash/pull/6023)), the changes needed were not extended to parallelized linting added in [dash#4637](https://github.com/dashpay/dash/pull/4637).

  This meant the match expression didn't include non-shell scripts and additionally, `parallel` interpreted all scripts as Bash scripts. This has been resolved in this PR.

* [bitcoin#23212](https://github.com/bitcoin/bitcoin/pull/23212) is partial as `--ignore-missing-imports` has not been removed from `mypy` arguments as the version of `mypy` used in the PR (0.910) isn't syntax aware to `imports` like [here](https://github.com/dashpay/dash/blob/f9d044d5ec7cc492d10fd4a89f85927b00305e83/test/functional/test_framework/crypto/bip324_cipher.py#L16-L17) and [here](https://github.com/dashpay/dash/blob/f9d044d5ec7cc492d10fd4a89f85927b00305e83/test/functional/test_framework/crypto/muhash.py#L9).

  <details>

  <summary>Lint errors:</summary>

  ```
  dash@96b217d539f7:/src/dash$ ./test/lint/lint-python.sh
  Consider install flake8-cached for cached flake8 results.
  [...]
  test/functional/test_framework/crypto/muhash.py:9: error: Skipping analyzing ".chacha20": found module but no type hints or library stubs  [import]
  test/functional/test_framework/crypto/ellswift.py:15: error: Cannot find implementation or library stub for module named "test_framework.crypto.secp256k1"  [import]
  test/functional/test_framework/crypto/bip324_cipher.py:16: error: Skipping analyzing ".chacha20": found module but no type hints or library stubs  [import]
  test/functional/test_framework/crypto/bip324_cipher.py:17: error: Skipping analyzing ".poly1305": found module but no type hints or library stubs  [import]
  test/functional/test_framework/messages.py:29: error: Cannot find implementation or library stub for module named "test_framework.crypto.siphash"  [import]
  test/functional/test_framework/messages.py:32: error: Cannot find implementation or library stub for module named "dash_hash"  [import]
  test/functional/test_framework/script.py:20: error: Cannot find implementation or library stub for module named "test_framework.crypto.ripemd160"  [import]
  test/functional/test_framework/key.py:16: error: Cannot find implementation or library stub for module named "test_framework.crypto"  [import]
  test/functional/test_framework/v2_p2p.py:9: error: Cannot find implementation or library stub for module named "test_framework.crypto.bip324_cipher"  [import]
  test/functional/test_framework/v2_p2p.py:10: error: Cannot find implementation or library stub for module named "test_framework.crypto.chacha20"  [import]
  test/functional/test_framework/v2_p2p.py:11: error: Cannot find implementation or library stub for module named "test_framework.crypto.ellswift"  [import]
  test/functional/test_framework/v2_p2p.py:12: error: Cannot find implementation or library stub for module named "test_framework.crypto.hkdf"  [import]
  test/functional/p2p_v2_encrypted.py:22: error: Cannot find implementation or library stub for module named "test_framework.crypto.chacha20"  [import]
  test/functional/feature_utxo_set_hash.py:12: error: Cannot find implementation or library stub for module named "test_framework.crypto.muhash"  [import]
  Found 17 errors in 12 files (checked 275 source files)
  ```

  </details>

  And upgrading to the latest version of `mypy` used upstream (1.4.1, [source](https://github.com/bitcoin/bitcoin/blob/f7144b24be09e7db2173a0b15d73f99a08b98118/ci/lint/04_install.sh#L52)) brings syntax awareness but new errors.

  <details>

  <summary>Lint errors:</summary>

  ```
  dash@96b217d539f7:/src/dash$ ./test/lint/lint-python.sh
  Consider install flake8-cached for cached flake8 results.
  [...]
  test/functional/test_framework/coverage.py:23: error: Incompatible default for argument "coverage_logfile" (default has type "None", argument has type "str")  [assignment]
  test/functional/test_framework/coverage.py:23: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
  test/functional/test_framework/coverage.py:23: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
  test/functional/test_framework/util.py:322: error: Incompatible default for argument "timeout" (default has type "None", argument has type "int")  [assignment]
  test/functional/test_framework/util.py:322: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
  test/functional/test_framework/util.py:322: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
  test/functional/test_framework/util.py:322: error: Incompatible default for argument "coveragedir" (default has type "None", argument has type "str")  [assignment]
  test/functional/test_framework/messages.py:32: error: Cannot find implementation or library stub for module named "dash_hash"  [import]
  test/functional/test_framework/test_framework.py:122: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
  test/functional/test_framework/test_framework.py:123: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
  test/functional/test_framework/test_framework.py:124: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
  test/functional/test_framework/test_framework.py:125: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
  test/functional/p2p_message_capture.py:48: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
  Found 6 errors in 5 files (checked 275 source files)
  ```

  </details>

* [bitcoin#23462](https://github.com/bitcoin/bitcoin/pull/23462) is partial as neither `SC2046` nor `SC2086` have been enabled. This backport was done for the sole purpose of backporting changes to shell scripts that would be replaced with their Python counterparts in later backports.

  The necessary changes needed to enable `SC2046` and `SC2086` will be done in a later pull request.

* `actions/checkout` does not do a full clone nor tracks the default branch by default. It is already documented that `git merge-base` (used by `lint-git-commit-check.py` and `lint-whitespace.py`, [source](https://github.com/dashpay/dash/blob/4aaa314a575a318683f3d15f16f396ac1a44a913/test/lint/lint-git-commit-check.py#L45)) doesn't play nice with it (see [actions/checkout#423](https://github.com/actions/checkout/discussions/423)).

  No such issue exists on GitLab ([build](https://gitlab.com/dashpay/dash/-/jobs/8456260939#L113)), but it remains present on GitHub Actions ([build](https://github.com/dashpay/dash/actions/runs/11996049800/job/33440106874?pr=6428#step:7:103)). This PR does a full clone, switches to the `develop` branch, then switches back to the intended commit, as a workaround (see working build [here](https://github.com/dashpay/dash/pull/6428#issuecomment-2495980410))

* Changes to `run-lint-format-strings.py` made in [dash#4807](https://github.com/dashpay/dash/pull/4807) were reverted as they were interfering with `lint-format-strings.py`

  <details>

  <summary>Lint error:</summary>

  ```
  dash@96b217d539f7:/src/dash$ ./test/lint/lint-format-strings.py
  Traceback (most recent call last):
    File "/src/dash/test/lint/run-lint-format-strings.py", line 308, in <module>
      main()
    File "/src/dash/test/lint/run-lint-format-strings.py", line 304, in main
      sys.exit(max(exit_codes))
  ValueError: max() arg is an empty sequence
  Traceback (most recent call last):
    File "/src/dash/test/lint/run-lint-format-strings.py", line 308, in <module>
      main()
    File "/src/dash/test/lint/run-lint-format-strings.py", line 304, in main
      sys.exit(max(exit_codes))
  ValueError: max() arg is an empty sequence
  ```

  </details>

## Breaking Changes

None expected.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas **(note: N/A)**
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Comment by kwvg** [comment]
GitHub Actions run for 4aaa314, https://github.com/kwvg/dash/actions/runs/11996049561.

**Review by knst — APPROVED** [review]
LGTM 4aaa314a575a318683f3d15f16f396ac1a44a913

**Comment by kwvg** [comment]
GitHub Actions run for 852f55e, https://github.com/kwvg/dash/actions/runs/12044056813.

File: test/lint/lint-whitespace.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                 "src/leveldb/",
+                 "src/crc32c/",
+                 "src/secp256k1/",
+                 "src/minisketch/",
+                 "src/univalue/",
+                 "src/dashbls/",
+                 "src/immer/",
+                 "src/util/expected.h",
+                 "doc/release-notes/",
+                 "src/qt/locale"]
```
Comment (UdjinM6) [inline test/lint/lint-whitespace.py:30]: 24844: note/nit: "src/qt/locale" was added in 16988 but that's a "translations update" PR (i.e. DNM) so it's probably ok to simply fix it here.

**Review by UdjinM6 — APPROVED** [review]
utACK 852f55e23cbe3be3851177883e5f54695291719b

(side note: there is no `mapfile` on macOS so some linters no longer work locally for me😞 )

File: test/lint/lint-whitespace.py (hunk 1 above)
Comment (kwvg) [reply test/lint/lint-whitespace.py:30]: Also, `mapfile` shouldn't be a problem after #6429, the only shell scripts that'll remain then are `commit-script-check.sh`, `git-subtree-check.sh` and `lint-cppcheck-dash.sh` (former two remain shell scripts upstream[1] and the latter is a Dash-specific script)

[1] - [`commit-script-check.sh`](https://github.com/bitcoin/bitcoin/blob/ff873a20a7f48698da3334acc570d97a07016ddf/test/lint/commit-script-check.sh), [`git-subtree-check.sh`](https://github.com/bitcoin/bitcoin/blob/ff873a20a7f48698da3334acc570d97a07016ddf/test/lint/git-subtree-check.sh)

**Pull request merged**